File: Scala/data_structures/src/main/scala-2.12/hashtables/RabinKarp.scala
package hashtables

object RabinKarp {
  import scala.annotation.tailrec
  import scala.util.Random
  import utils.RichNumber._

  private val rng = new Random(2112L)

  private[hashtables] val Prime: Long = 1000000007

  private def polynomialHashing(string: String, prime: Long, x: Long): Int =
    string.foldRight(0L){ case (char, acc) => (acc * x + char.toInt) % prime }.toInt

  private def calcPowerOfX(x: Long, prime: Long, exponent: Int): Long =
    (0 until exponent).foldLeft(1L){ case (acc, _) => (acc * x) % prime }

  private def substringHashes(text: String, textLength: Int, patternLength: Int, prime: Long, x: Long): Vector[Int] = {
    val length: Int = textLength - patternLength + 1
    val hashValues: Array[Int] = Array.fill(length)(0)
    val lastSubstring: String = text.substring(textLength - patternLength)
    hashValues(length - 1) = polynomialHashing(lastSubstring, prime, x)
    val xPower: Long = calcPowerOfX(x, prime, patternLength)
    for { ix <- length - 2 to 0 by -1 } {
      val hashed: Long = x * hashValues(ix + 1) + text(ix).toInt - xPower * text(ix + patternLength).toInt
      hashValues(ix) = hashed.mod(prime).toInt
    }
    hashValues.toVector
  }

  private def getMatchingIndices(text: String,
                                 textLength: Int,
                                 pattern: String,
                                 patternLength: Int,
                                 patternHash: Int,
                                 hashValues: Vector[Int]): List[Int] = {
    val length: Int = textLength - patternLength + 1
    @tailrec
    def loop(ix: Int, acc: List[Int]): List[Int] = {
      if (ix == length) acc.reverse
      else if (patternHash == hashValues(ix) && pattern == text.substring(ix, ix + patternLength)) {
        loop(ix + 1, ix :: acc)
      } else loop(ix + 1, acc)
    }
    loop(0, Nil)
  }

  def findPatternInText(text: String, pattern: String, prime: Long): List[Int] = {
    val x: Long = rng.nextInt(prime.toInt - 2) + 1
    val patternLength: Int = pattern.length
    val textLength: Int = text.length
    val patternHash: Int = polynomialHashing(pattern, prime, x)
    val hashValues: Vector[Int] = substringHashes(text, textLength, patternLength, prime, x)
    getMatchingIndices(text, textLength, pattern, patternLength, patternHash, hashValues)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val pattern: String = reader.next()
    val text: String = reader.next()
    val indices: List[Int] = findPatternInText(text, pattern, Prime)
    println(indices.mkString(" "))
  }
}

File: OCaml/data_structures/hashtables/rabin_karp.ml
open Batteries

let modulo a p =
    let result = a mod p in
    if result >= 0 then result
    else result + p


let polynomial_hashing s prime x =
    String.fold_right (fun chr acc -> (acc * x + Char.code chr) mod prime) s 0


let calc_power_of_x x prime exponent =
    let rec loop acc k =
        if k = exponent then acc
        else loop ((acc * x) mod prime) (k + 1)
    in loop 1 0


let calc_substring_hashes text text_length pattern_length prime x =
    let length = text_length - pattern_length + 1 in
    let hash_values = Array.make length 0 in
    let last_substring = String.slice ~first:(length - 1) text in
    hash_values.(length - 1) <- polynomial_hashing last_substring prime x;
    let x_power = calc_power_of_x x prime pattern_length in
    let code_at_position str k = Char.code (String.get str k) in
    for ix = (length - 2) downto 0 do
        let hashed = x * hash_values.(ix + 1) +
            code_at_position text ix - x_power * code_at_position text (ix + pattern_length) in
        hash_values.(ix) <- modulo hashed prime
    done;
    hash_values


let get_matching_indices text text_length pattern pattern_length pattern_hash substring_hashes =
    let length = text_length - pattern_length + 1 in
    let rec loop acc ix =
        if ix = length then List.rev acc
        else if pattern_hash = substring_hashes.(ix) &&
                pattern = String.slice ~first:ix ~last:(ix + pattern_length) text then loop (ix :: acc) (ix + 1)
        else loop acc (ix + 1)
    in loop [] 0


let find_pattern_in_text text pattern prime =
    let () = Random.init 2112 in
    let x = 1 + Random.int (prime - 2) in
    let pattern_length = String.length pattern in
    let text_length = String.length text in
    let pattern_hash_value = polynomial_hashing pattern prime x in
    let substring_hash_values = calc_substring_hashes text text_length pattern_length prime x
    in get_matching_indices text text_length pattern pattern_length pattern_hash_value substring_hash_values


let () =
    let pattern = read_line() in
    let text = read_line () in
    let prime = 1000000007 in
    let indices = find_pattern_in_text text pattern prime in
    List.iter (function elem -> print_int elem; print_string " ") indices;
    print_newline()
